controllers/drain_controller.go

```diff
@@ -0,0 +1,463 @@
+
+	//TODO: change this to save it on runtime
+	mcpPauseDrainNodeList := &corev1.NodeList{}
+	err = dr.List(ctx, mcpPauseDrainNodeList, &client.ListOptions{LabelSelector: dr.mcpPauseDrainSelector})
+	if err != nil {
+		// Failed to get node list
+		reqLogger.Error(err, "Error occurred on LIST nodes request from API server")
+		return reconcile.Result{}, err
+	}
+
+	DrainingNodeList := &corev1.NodeList{}
+	err = dr.List(ctx, DrainingNodeList, &client.ListOptions{LabelSelector: dr.mcpPauseDrainSelector})
+	if err != nil {
+		// Failed to get node list
+		reqLogger.Error(err, "Error occurred on LIST nodes request from API server")
+		return reconcile.Result{}, err
+	}
+
+	reqLogger.Info("Max node allowed to be draining at the same time", "MaxParallelNodeConfiguration", maxParallelNodeConfiguration)
+	drainingNodes := len(mcpPauseDrainNodeList.Items) + len(DrainingNodeList.Items)
+	reqLogger.Info("Count of draining", "drainingNodes", drainingNodes)
+
+	// if both are Idle we don't need to do anything
+	if nodeDrainAnnotation == constants.DrainIdle &&
+		nodeStateDrainAnnotationCurrent == constants.DrainIdle {
+		reqLogger.Info("node and nodeState are on idle nothing todo")
+		return reconcile.Result{}, nil
+	}
+
+	if nodeDrainAnnotation == constants.DrainIdle &&
+		nodeStateDrainAnnotationCurrent == constants.DrainComplete {
+		err = dr.completeDrain(ctx, node)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainIdle, dr.Client)
+		return ctrl.Result{}, err
+	}
+
+	if drainingNodes >= maxParallelNodeConfiguration {
+		reqLogger.Info("MaxParallelNodeConfiguration limit reached for draining nodes re-enqueue the request")
+		// TODO: make this time configurable
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	// the node request to drain, but we are at the limit so we re-enqueue the request
+	if nodeDrainAnnotation == constants.DrainRequired &&
+		nodeStateDrainAnnotationCurrent == constants.DrainIdle {
+		// pause MCP if we are on openshift but not hypershift deployment method as there is MCP in hypershift
+		if !dr.openshiftContext.IsOpenshiftCluster() || (dr.openshiftContext.IsOpenshiftCluster() && dr.openshiftContext.IsHypershift()) {
+			err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.Draining, dr.Client)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			nodeStateDrainAnnotationCurrent = constants.Draining
+		} else if dr.openshiftContext.IsOpenshiftCluster() && !dr.openshiftContext.IsHypershift() {
+			nodePoolName, err := dr.openshiftContext.GetNodeMachinePoolName(node)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if err := dr.pauseMCP(nodeNetworkState, nodePoolName); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			nodeStateDrainAnnotationCurrent = nodeNetworkState.Annotations[constants.NodeStateDrainAnnotationCurrent]
```

Comment:
This value is always `constants.DrainIdle`, as the `if` contains `nodeStateDrainAnnotationCurrent == constants.DrainIdle` and that value comes from L139. isn't it?

Reply:
I just noticed `nodeNetworkState.Annotations` is changed to `Draining_MCP_Paused` by `pauseMCP()`

Reply:
right, I don't like it but this was how it was working I just copy the function from the daemon.

In the parallel PR I will change that but in this one I am trying to reduce the amount of changes